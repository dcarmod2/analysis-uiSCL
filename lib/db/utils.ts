import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
