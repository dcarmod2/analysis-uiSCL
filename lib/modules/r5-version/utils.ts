import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
