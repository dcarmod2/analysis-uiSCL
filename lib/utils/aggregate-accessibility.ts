import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
