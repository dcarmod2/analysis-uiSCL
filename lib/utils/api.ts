// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
