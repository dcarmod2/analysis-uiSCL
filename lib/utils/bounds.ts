import lonlat from '@conveyal/lonlat'

import L from 'lib/leaflet'
import message from 'lib/message'

import reprojectCoordinates from './reproject-coordinates'

const ANTIMERIDIAN_WARNING =
  'We do not currently support bounds that cross the antimeridian.'

type Bounds = {
  north: number
  south: number
  east: number
  west: number
  toString: () => string
}

export function toLatLngBounds(
  bounds: Bounds | L.LatLngBounds
): L.LatLngBounds {
  // Check if it's already a leaflet bounds object
  if (bounds instanceof L.LatLngBounds) return bounds
  return new L.LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds(bounds: L.LatLngBounds) {
  // Don't change the original bounds object
  let b = new L.LatLngBounds(bounds.getSouthWest(), bounds.getNorthEast())

  if (b.getWest() < -180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), -180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), 180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds(bounds) {
  return new L.LatLngBounds(
    lonlat.toLeaflet(reprojectCoordinates(bounds.getSouthWest())),
    lonlat.toLeaflet(reprojectCoordinates(bounds.getNorthEast()))
  )
}
