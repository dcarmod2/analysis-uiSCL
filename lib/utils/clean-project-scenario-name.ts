//
import message from 'lib/message'

import {UNDEFINED_PROJECT_NAME} from '../constants'

export default function cleanProjectScenarioName(project, scenarioIndex) {
  if (!project) {
    return UNDEFINED_PROJECT_NAME
  }
  const scenarioName =
    project.variants[scenarioIndex] || message('variant.baseline')
  const description = `${project.name}-${scenarioName}`
  return description.replace(/[^a-zA-Z0-9]/, '-')
}
