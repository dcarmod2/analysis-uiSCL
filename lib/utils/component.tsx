import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
