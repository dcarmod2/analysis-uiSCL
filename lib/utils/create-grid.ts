/**
 * Create a Grid from an ArrayBuffer retrieved from the server or S3. Binary
 * data that uses a common header format used across Conveyal libraries.
 */
export default function create(data: ArrayBuffer): CL.ParsedGrid {
  const array = new Int32Array(data, 4 * 5)
  const header = new Int32Array(data)

  let min = Infinity
  let max = -Infinity

  for (let i = 0, prev = 0; i < array.length; i++) {
    array[i] = prev += array[i] // de-delta code
    if (prev < min) min = prev
    if (prev > max) max = prev
  }

  const width = header[3]
  const height = header[4]

  function contains(x: number, y: number) {
    return x >= 0 && x < width && y >= 0 && y < height
  }

  // Return 0 for cells outside of the grid
  function getValue(x: number, y: number) {
    if (contains(x, y)) return array[y * width + x]
    return 0
  }

  // parse header
  return {
    zoom: header[0],
    west: header[1],
    north: header[2],
    width,
    height,
    data: array,
    min,
    max,
    contains,
    getValue
  }
}
