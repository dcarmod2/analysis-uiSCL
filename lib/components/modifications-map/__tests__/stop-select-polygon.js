import React from 'react'
import {Map} from 'react-leaflet'

import {testAndSnapshot} from 'lib/utils/component'

import StopSelectPolygon from '../stop-select-polygon'

const props = {
  routeStops: [
    {
      stop_id: '1',
      stop_lat: 12,
      stop_lon: 34,
      stop_name: 'mock gtfs stop'
    }
  ],
  selectStops: jest.fn()
}

function StopSelectPolygonTest(props) {
  return (
    <Map>
      <StopSelectPolygon {...props} />
    </Map>
  )
}

testAndSnapshot(StopSelectPolygonTest, props)
