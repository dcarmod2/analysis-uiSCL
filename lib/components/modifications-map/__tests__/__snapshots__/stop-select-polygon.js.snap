// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component StopSelectPolygonTest snapshot(mount) 1`] = `
<StopSelectPolygonTest
  routeStops={
    Array [
      Object {
        "stop_id": "1",
        "stop_lat": 12,
        "stop_lon": 34,
        "stop_name": "mock gtfs stop",
      },
    ]
  }
  selectStops={[MockFunction]}
>
  <Map>
    <div>
      <StopSelectPolygon
        routeStops={
          Array [
            Object {
              "stop_id": "1",
              "stop_lat": 12,
              "stop_lon": 34,
              "stop_name": "mock gtfs stop",
            },
          ]
        }
        selectStops={[MockFunction]}
      >
        <DrawPolygon
          activateOnMount={true}
          onPolygon={[Function]}
        >
          <ForwardRef(Leaflet(FeatureGroup))>
            <FeatureGroup
              leaflet={
                Object {
                  "layerContainer": MapMock {
                    "_addLayers": [MockFunction],
                    "_addZoomLimit": [MockFunction],
                    "_animMoveEnd": [MockFunction],
                    "_animateZoom": [MockFunction],
                    "_catchTransitionEnd": [MockFunction],
                    "_checkIfLoaded": [MockFunction],
                    "_clearControlPos": [MockFunction],
                    "_clearHandlers": [MockFunction],
                    "_container": <div />,
                    "_createAnimProxy": [MockFunction],
                    "_createRenderer": [MockFunction],
                    "_destroyAnimProxy": [MockFunction],
                    "_draggableMoved": [MockFunction],
                    "_events": Object {
                      "move": Array [
                        Object {
                          "ctx": undefined,
                          "fn": [Function],
                        },
                      ],
                      "moveend": Array [
                        Object {
                          "ctx": undefined,
                          "fn": [Function],
                        },
                      ],
                    },
                    "_findEventTargets": [MockFunction],
                    "_fireDOMEvent": [MockFunction],
                    "_getBoundsCenterZoom": [MockFunction],
                    "_getBoundsOffset": [MockFunction],
                    "_getCenterLayerPoint": [MockFunction],
                    "_getCenterOffset": [MockFunction],
                    "_getMapPanePos": [MockFunction],
                    "_getNewPixelOrigin": [MockFunction],
                    "_getPaneRenderer": [MockFunction],
                    "_getTopLeftPoint": [MockFunction],
                    "_getZoomSpan": [MockFunction],
                    "_handleDOMEvent": [MockFunction],
                    "_handleGeolocationError": [MockFunction],
                    "_handleGeolocationResponse": [MockFunction],
                    "_initContainer": [MockFunction],
                    "_initControlPos": [MockFunction],
                    "_initEvents": [MockFunction],
                    "_initLayout": [MockFunction],
                    "_initPanes": [MockFunction],
                    "_latLngBoundsToNewLayerBounds": [MockFunction],
                    "_latLngToNewLayerPoint": [MockFunction],
                    "_limitCenter": [MockFunction],
                    "_limitOffset": [MockFunction],
                    "_limitZoom": [MockFunction],
                    "_move": [MockFunction],
                    "_moveEnd": [MockFunction],
                    "_moveStart": [MockFunction],
                    "_moved": [MockFunction],
                    "_nothingToAnimate": [MockFunction],
                    "_off": [Function],
                    "_on": [Function],
                    "_onMoveEnd": [MockFunction],
                    "_onPanTransitionEnd": [MockFunction],
                    "_onPanTransitionStep": [MockFunction],
                    "_onResize": [MockFunction],
                    "_onScroll": [MockFunction],
                    "_onZoomTransitionEnd": [MockFunction],
                    "_panInsideMaxBounds": [MockFunction],
                    "_propagateEvent": [Function],
                    "_rawPanBy": [MockFunction],
                    "_rebound": [MockFunction],
                    "_removeZoomLimit": [MockFunction],
                    "_resetView": [MockFunction],
                    "_stop": [MockFunction],
                    "_tryAnimatedPan": [MockFunction],
                    "_tryAnimatedZoom": [MockFunction],
                    "_updateZoomLevels": [MockFunction],
                    "addControl": [MockFunction],
                    "addEventListener": [Function],
                    "addEventParent": [Function],
                    "addHandler": [MockFunction],
                    "addLayer": [MockFunction] {
                      "calls": Array [
                        Array [
                          NewClass {},
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "addOneTimeEventListener": [Function],
                    "callInitHooks": [MockFunction],
                    "clearAllEventListeners": [Function],
                    "closePopup": [MockFunction],
                    "closeTooltip": [MockFunction],
                    "containerPointToLatLng": [MockFunction],
                    "containerPointToLayerPoint": [MockFunction],
                    "createPane": [MockFunction],
                    "distance": [MockFunction],
                    "eachLayer": [MockFunction],
                    "fire": [Function],
                    "fireEvent": [Function],
                    "fitBounds": [MockFunction],
                    "fitWorld": [MockFunction],
                    "flyTo": [MockFunction],
                    "flyToBounds": [MockFunction],
                    "getBounds": [MockFunction],
                    "getBoundsZoom": [MockFunction],
                    "getCenter": [MockFunction],
                    "getContainer": [MockFunction],
                    "getMaxZoom": [MockFunction],
                    "getMinZoom": [MockFunction],
                    "getPane": [MockFunction],
                    "getPanes": [MockFunction],
                    "getPixelBounds": [MockFunction],
                    "getPixelOrigin": [MockFunction],
                    "getPixelWorldBounds": [MockFunction],
                    "getRenderer": [MockFunction],
                    "getScaleZoom": [MockFunction],
                    "getSize": [MockFunction],
                    "getZoom": [MockFunction],
                    "getZoomScale": [MockFunction],
                    "hasEventListeners": [Function],
                    "hasLayer": [MockFunction],
                    "initialize": [MockFunction],
                    "invalidateSize": [MockFunction],
                    "latLngToContainerPoint": [MockFunction],
                    "latLngToLayerPoint": [MockFunction],
                    "layerPointToContainerPoint": [MockFunction],
                    "layerPointToLatLng": [MockFunction],
                    "listens": [Function],
                    "locate": [MockFunction],
                    "mouseEventToContainerPoint": [MockFunction],
                    "mouseEventToLatLng": [MockFunction],
                    "mouseEventToLayerPoint": [MockFunction],
                    "off": [Function],
                    "on": [Function],
                    "once": [Function],
                    "openPopup": [MockFunction],
                    "openTooltip": [MockFunction],
                    "options": Object {
                      "attributionControl": true,
                      "bounceAtZoomLimits": true,
                      "boxZoom": true,
                      "center": undefined,
                      "closePopupOnClick": true,
                      "crs": Object {
                        "R": 6371000,
                        "code": "EPSG:3857",
                        "distance": [Function],
                        "getProjectedBounds": [Function],
                        "infinite": false,
                        "latLngToPoint": [Function],
                        "pointToLatLng": [Function],
                        "project": [Function],
                        "projection": Object {
                          "MAX_LATITUDE": 85.0511287798,
                          "R": 6378137,
                          "bounds": Object {
                            "max": Object {
                              "x": 20037508.342789244,
                              "y": 20037508.342789244,
                            },
                            "min": Object {
                              "x": -20037508.342789244,
                              "y": -20037508.342789244,
                            },
                          },
                          "project": [Function],
                          "unproject": [Function],
                        },
                        "scale": [Function],
                        "transformation": Object {
                          "_a": 2.495320233665337e-8,
                          "_b": 0.5,
                          "_c": -2.495320233665337e-8,
                          "_d": 0.5,
                        },
                        "unproject": [Function],
                        "wrapLatLng": [Function],
                        "wrapLatLngBounds": [Function],
                        "wrapLng": Array [
                          -180,
                          180,
                        ],
                        "zoom": [Function],
                      },
                      "doubleClickZoom": true,
                      "dragging": true,
                      "easeLinearity": 0.2,
                      "fadeAnimation": true,
                      "inertia": true,
                      "inertiaDeceleration": 3400,
                      "inertiaMaxSpeed": Infinity,
                      "keyboard": true,
                      "keyboardPanDelta": 80,
                      "layers": Array [],
                      "markerZoomAnimation": true,
                      "maxBounds": undefined,
                      "maxBoundsViscosity": 0,
                      "maxZoom": undefined,
                      "minZoom": undefined,
                      "renderer": undefined,
                      "scrollWheelZoom": true,
                      "tap": true,
                      "tapTolerance": 15,
                      "touchZoom": undefined,
                      "trackResize": true,
                      "transform3DLimit": 8388608,
                      "wheelDebounceTime": 40,
                      "wheelPxPerZoomLevel": 60,
                      "worldCopyJump": false,
                      "zoom": undefined,
                      "zoomAnimation": true,
                      "zoomAnimationThreshold": 4,
                      "zoomControl": true,
                      "zoomDelta": 1,
                      "zoomSnap": 1,
                    },
                    "panBy": [MockFunction],
                    "panInside": [MockFunction],
                    "panInsideBounds": [MockFunction],
                    "panTo": [MockFunction],
                    "project": [MockFunction],
                    "remove": [MockFunction],
                    "removeControl": [MockFunction],
                    "removeEventListener": [Function],
                    "removeEventParent": [Function],
                    "removeLayer": [MockFunction],
                    "setMaxBounds": [MockFunction],
                    "setMaxZoom": [MockFunction],
                    "setMinZoom": [MockFunction],
                    "setView": [MockFunction],
                    "setZoom": [MockFunction],
                    "setZoomAround": [MockFunction],
                    "stop": [MockFunction],
                    "stopLocate": [MockFunction],
                    "unproject": [MockFunction],
                    "whenReady": [MockFunction],
                    "wrapLatLng": [MockFunction],
                    "wrapLatLngBounds": [MockFunction],
                    "zoomIn": [MockFunction],
                    "zoomOut": [MockFunction],
                  },
                  "map": MapMock {
                    "_addLayers": [MockFunction],
                    "_addZoomLimit": [MockFunction],
                    "_animMoveEnd": [MockFunction],
                    "_animateZoom": [MockFunction],
                    "_catchTransitionEnd": [MockFunction],
                    "_checkIfLoaded": [MockFunction],
                    "_clearControlPos": [MockFunction],
                    "_clearHandlers": [MockFunction],
                    "_container": <div />,
                    "_createAnimProxy": [MockFunction],
                    "_createRenderer": [MockFunction],
                    "_destroyAnimProxy": [MockFunction],
                    "_draggableMoved": [MockFunction],
                    "_events": Object {
                      "move": Array [
                        Object {
                          "ctx": undefined,
                          "fn": [Function],
                        },
                      ],
                      "moveend": Array [
                        Object {
                          "ctx": undefined,
                          "fn": [Function],
                        },
                      ],
                    },
                    "_findEventTargets": [MockFunction],
                    "_fireDOMEvent": [MockFunction],
                    "_getBoundsCenterZoom": [MockFunction],
                    "_getBoundsOffset": [MockFunction],
                    "_getCenterLayerPoint": [MockFunction],
                    "_getCenterOffset": [MockFunction],
                    "_getMapPanePos": [MockFunction],
                    "_getNewPixelOrigin": [MockFunction],
                    "_getPaneRenderer": [MockFunction],
                    "_getTopLeftPoint": [MockFunction],
                    "_getZoomSpan": [MockFunction],
                    "_handleDOMEvent": [MockFunction],
                    "_handleGeolocationError": [MockFunction],
                    "_handleGeolocationResponse": [MockFunction],
                    "_initContainer": [MockFunction],
                    "_initControlPos": [MockFunction],
                    "_initEvents": [MockFunction],
                    "_initLayout": [MockFunction],
                    "_initPanes": [MockFunction],
                    "_latLngBoundsToNewLayerBounds": [MockFunction],
                    "_latLngToNewLayerPoint": [MockFunction],
                    "_limitCenter": [MockFunction],
                    "_limitOffset": [MockFunction],
                    "_limitZoom": [MockFunction],
                    "_move": [MockFunction],
                    "_moveEnd": [MockFunction],
                    "_moveStart": [MockFunction],
                    "_moved": [MockFunction],
                    "_nothingToAnimate": [MockFunction],
                    "_off": [Function],
                    "_on": [Function],
                    "_onMoveEnd": [MockFunction],
                    "_onPanTransitionEnd": [MockFunction],
                    "_onPanTransitionStep": [MockFunction],
                    "_onResize": [MockFunction],
                    "_onScroll": [MockFunction],
                    "_onZoomTransitionEnd": [MockFunction],
                    "_panInsideMaxBounds": [MockFunction],
                    "_propagateEvent": [Function],
                    "_rawPanBy": [MockFunction],
                    "_rebound": [MockFunction],
                    "_removeZoomLimit": [MockFunction],
                    "_resetView": [MockFunction],
                    "_stop": [MockFunction],
                    "_tryAnimatedPan": [MockFunction],
                    "_tryAnimatedZoom": [MockFunction],
                    "_updateZoomLevels": [MockFunction],
                    "addControl": [MockFunction],
                    "addEventListener": [Function],
                    "addEventParent": [Function],
                    "addHandler": [MockFunction],
                    "addLayer": [MockFunction] {
                      "calls": Array [
                        Array [
                          NewClass {},
                        ],
                      ],
                      "results": Array [
                        Object {
                          "type": "return",
                          "value": undefined,
                        },
                      ],
                    },
                    "addOneTimeEventListener": [Function],
                    "callInitHooks": [MockFunction],
                    "clearAllEventListeners": [Function],
                    "closePopup": [MockFunction],
                    "closeTooltip": [MockFunction],
                    "containerPointToLatLng": [MockFunction],
                    "containerPointToLayerPoint": [MockFunction],
                    "createPane": [MockFunction],
                    "distance": [MockFunction],
                    "eachLayer": [MockFunction],
                    "fire": [Function],
                    "fireEvent": [Function],
                    "fitBounds": [MockFunction],
                    "fitWorld": [MockFunction],
                    "flyTo": [MockFunction],
                    "flyToBounds": [MockFunction],
                    "getBounds": [MockFunction],
                    "getBoundsZoom": [MockFunction],
                    "getCenter": [MockFunction],
                    "getContainer": [MockFunction],
                    "getMaxZoom": [MockFunction],
                    "getMinZoom": [MockFunction],
                    "getPane": [MockFunction],
                    "getPanes": [MockFunction],
                    "getPixelBounds": [MockFunction],
                    "getPixelOrigin": [MockFunction],
                    "getPixelWorldBounds": [MockFunction],
                    "getRenderer": [MockFunction],
                    "getScaleZoom": [MockFunction],
                    "getSize": [MockFunction],
                    "getZoom": [MockFunction],
                    "getZoomScale": [MockFunction],
                    "hasEventListeners": [Function],
                    "hasLayer": [MockFunction],
                    "initialize": [MockFunction],
                    "invalidateSize": [MockFunction],
                    "latLngToContainerPoint": [MockFunction],
                    "latLngToLayerPoint": [MockFunction],
                    "layerPointToContainerPoint": [MockFunction],
                    "layerPointToLatLng": [MockFunction],
                    "listens": [Function],
                    "locate": [MockFunction],
                    "mouseEventToContainerPoint": [MockFunction],
                    "mouseEventToLatLng": [MockFunction],
                    "mouseEventToLayerPoint": [MockFunction],
                    "off": [Function],
                    "on": [Function],
                    "once": [Function],
                    "openPopup": [MockFunction],
                    "openTooltip": [MockFunction],
                    "options": Object {
                      "attributionControl": true,
                      "bounceAtZoomLimits": true,
                      "boxZoom": true,
                      "center": undefined,
                      "closePopupOnClick": true,
                      "crs": Object {
                        "R": 6371000,
                        "code": "EPSG:3857",
                        "distance": [Function],
                        "getProjectedBounds": [Function],
                        "infinite": false,
                        "latLngToPoint": [Function],
                        "pointToLatLng": [Function],
                        "project": [Function],
                        "projection": Object {
                          "MAX_LATITUDE": 85.0511287798,
                          "R": 6378137,
                          "bounds": Object {
                            "max": Object {
                              "x": 20037508.342789244,
                              "y": 20037508.342789244,
                            },
                            "min": Object {
                              "x": -20037508.342789244,
                              "y": -20037508.342789244,
                            },
                          },
                          "project": [Function],
                          "unproject": [Function],
                        },
                        "scale": [Function],
                        "transformation": Object {
                          "_a": 2.495320233665337e-8,
                          "_b": 0.5,
                          "_c": -2.495320233665337e-8,
                          "_d": 0.5,
                        },
                        "unproject": [Function],
                        "wrapLatLng": [Function],
                        "wrapLatLngBounds": [Function],
                        "wrapLng": Array [
                          -180,
                          180,
                        ],
                        "zoom": [Function],
                      },
                      "doubleClickZoom": true,
                      "dragging": true,
                      "easeLinearity": 0.2,
                      "fadeAnimation": true,
                      "inertia": true,
                      "inertiaDeceleration": 3400,
                      "inertiaMaxSpeed": Infinity,
                      "keyboard": true,
                      "keyboardPanDelta": 80,
                      "layers": Array [],
                      "markerZoomAnimation": true,
                      "maxBounds": undefined,
                      "maxBoundsViscosity": 0,
                      "maxZoom": undefined,
                      "minZoom": undefined,
                      "renderer": undefined,
                      "scrollWheelZoom": true,
                      "tap": true,
                      "tapTolerance": 15,
                      "touchZoom": undefined,
                      "trackResize": true,
                      "transform3DLimit": 8388608,
                      "wheelDebounceTime": 40,
                      "wheelPxPerZoomLevel": 60,
                      "worldCopyJump": false,
                      "zoom": undefined,
                      "zoomAnimation": true,
                      "zoomAnimationThreshold": 4,
                      "zoomControl": true,
                      "zoomDelta": 1,
                      "zoomSnap": 1,
                    },
                    "panBy": [MockFunction],
                    "panInside": [MockFunction],
                    "panInsideBounds": [MockFunction],
                    "panTo": [MockFunction],
                    "project": [MockFunction],
                    "remove": [MockFunction],
                    "removeControl": [MockFunction],
                    "removeEventListener": [Function],
                    "removeEventParent": [Function],
                    "removeLayer": [MockFunction],
                    "setMaxBounds": [MockFunction],
                    "setMaxZoom": [MockFunction],
                    "setMinZoom": [MockFunction],
                    "setView": [MockFunction],
                    "setZoom": [MockFunction],
                    "setZoomAround": [MockFunction],
                    "stop": [MockFunction],
                    "stopLocate": [MockFunction],
                    "unproject": [MockFunction],
                    "whenReady": [MockFunction],
                    "wrapLatLng": [MockFunction],
                    "wrapLatLngBounds": [MockFunction],
                    "zoomIn": [MockFunction],
                    "zoomOut": [MockFunction],
                  },
                }
              }
            >
              <Component
                draw={
                  Object {
                    "circle": false,
                    "circlemarker": false,
                    "marker": false,
                    "polygon": true,
                    "polyline": false,
                    "rectangle": false,
                  }
                }
                onCreated={[Function]}
                position="bottomright"
              />
            </FeatureGroup>
          </ForwardRef(Leaflet(FeatureGroup))>
        </DrawPolygon>
      </StopSelectPolygon>
    </div>
  </Map>
</StopSelectPolygonTest>
`;
